### PR TITLE
JAMES-3467 Experimental cache for DomainList calls

### PR DESCRIPTION
**Pull request by chibenwa**
Upon my load testing campaign, I noticed `SELECT * FROM domains WHERE domain=xxx;` to be one of our "hot" query, both on JMAP load when JWT is used (because there is a lookup to the domain before granting access), but also backend processing.

This query load have a real impact (low but real) on the load we are able to handle.

On my use case (on premise in customer site) the domains (almost) never change, and, say, 10s propagation upon creation are acceptable.

Thus I would like to propose an opt-in cache mechanism to reduce the query load on Cassandra (but other implementation as well!). A cache on each node is ok, as the volume of domain is expected-to-be very low.

Of course, it should be explicitly configured, and disabled by default. Of course we should warn this is experimental.

Expected outcome is a 99% decrease of that query volume.

### CODE REVIEW

File: server/data/data-library/src/main/java/org/apache/james/domainlist/lib/AbstractDomainList.java [hunk 1]
```diff
@@ -199,15 +192,15 @@ public boolean containsDomain(Domain domain) throws DomainListException {
     }
 
     private ImmutableList<Domain> detectIps(Collection<Domain> domains) {
-        if (autoDetectIP) {
+        if (configuration.isAutoDetect() ) {
```
Comment (Arsnael) [inline server/data/data-library/src/main/java/org/apache/james/domainlist/lib/AbstractDomainList.java]: `isAutoDetectIp` ?

File: src/site/xdoc/server/config-domainlist.xml [hunk 2]
```diff
@@ -47,6 +47,17 @@ If autodetect is false, James will use only the specified domainnames. Defaults
...
+
+        <dt><strong>read.cache.enable</strong></dt>
+        <dd>Experimental. Boolean, defaults to false.
+            Whether or not to cache domainlist.contains calls. Enable a faster execution however writes will take time
+            to propagate.</dd>
+
+        <dt><strong>read.cache.expiracy</strong></dt>
+        <dd>Experimental. String (duration), defaults to 10 seconds (10s). Supported units are ms, s, m, h, d, w, month, y.
+            Expiracy of the cache. Longer means less reads are performed to the backend but writes will take longer to propagate.
+            Low values (a few seconds) are adivsed.</dd>
```
Comment (Arsnael) [inline src/site/xdoc/server/config-domainlist.xml:59]: s/adivsed/advised

**Review by Arsnael — APPROVED** [review]

**Comment by chibenwa** [comment]
test this please

File: server/data/data-cassandra/src/test/java/org/apache/james/domainlist/cassandra/CacheDomainListTest.java [hunk 3]
```diff
@@ -0,0 +1,115 @@
...
+        domainList.containsDomain(DOMAIN_1);
+
+        domainList.addDomain(DOMAIN_1);
+
+        domainList.getDomains();
+
+        assertThat(domainList.containsDomain(DOMAIN_1)).isEqualTo(true);
+    }
+
+    DNSService getDNSServer(final String hostName) throws UnknownHostException {
```
Comment (LanKhuat) [inline server/data/data-cassandra/src/test/java/org/apache/james/domainlist/cassandra/CacheDomainListTest.java:110]: private?

**Review by Arsnael — APPROVED** [review]

**Comment by chibenwa** [comment]
test this please

**Comment by chibenwa** [comment]
test this please

**Review by LanKhuat — APPROVED** [review]

**Review by Arsnael — APPROVED** [review]

**Comment by chibenwa** [comment]
Merged

**Pull request closed without merging**